File: 0008 - String to Integer (atoi)/solution.cpp
class Solution {
private:
    bool check(int sign, int res, int d){
        if(sign == 1){
            if(res > INT_MAX/10) return true;
            else if(res==INT_MAX/10 && d>7) return true;
        }
        else{
            if(res < INT_MIN/10) return true;
            else if(res==INT_MIN/10 && d>8) return true;
        }
        return false;
    }
public:
    int myAtoi(string s) {
        int res = 0, sign = 1, flag = 0;
        for(auto c : s){
            if(c==' ' && flag==0) continue;
            else if((c=='+' || c=='-') && flag==0) sign = c=='+'?1:-1, flag = 1;
            else if(c>='0' && c<='9'){
                flag = 1;
                if(check(sign, res, c-'0')){
                    res = sign==1?INT_MAX:INT_MIN;
                    break;
                }
                else res = res*10+sign*(c-'0');
            }
            else break;
        }
        return res;
    }
};

File: 0008 - String to Integer (atoi)/solution.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.lstrip()
        if not s:
            return 0
        res, MAX, MIN = 0, 2**31-1, -2**31
        sign = -1 if s[0]=='-' else 1
        i = 1 if s[0] in '+-' else 0
        while i<len(s) and s[i].isdigit():
            if res>MAX//10 or (res==MAX//10 and s[i]>'7'):
                return MAX if sign==1 else MIN
            res = 10*res+(ord(s[i])-ord('0')); i += 1
        return sign*res
        